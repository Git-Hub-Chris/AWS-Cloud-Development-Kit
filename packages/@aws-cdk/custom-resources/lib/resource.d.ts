import { cloudformation } from '@aws-cdk/aws-cloudformation';
import lambda = require('@aws-cdk/aws-lambda');
import sns = require('@aws-cdk/aws-sns');
import cdk = require('@aws-cdk/cdk');
/**
 * Collection of arbitrary properties
 */
export declare type Properties = {
    [key: string]: any;
};
/**
 * Properties to provide a Lambda-backed custom resource
 */
export interface CustomResourceProps {
    /**
     * The Lambda provider that implements this custom resource.
     *
     * We recommend using a SingletonLambda for this.
     *
     * Optional, exactly one of lamdaProvider or topicProvider must be set.
     */
    lambdaProvider?: lambda.LambdaRef;
    /**
     * The SNS Topic for the provider that implements this custom resource.
     *
     * Optional, exactly one of lamdaProvider or topicProvider must be set.
     */
    topicProvider?: sns.TopicRef;
    /**
     * Properties to pass to the Lambda
     */
    properties?: Properties;
}
/**
 * Custom resource that is implemented using a Lambda
 *
 * As a custom resource author, you should be publishing a subclass of this class
 * that hides the choice of provider, and accepts a strongly-typed properties
 * object with the properties your provider accepts.
 */
export declare class CustomResource extends cloudformation.CustomResource {
    private readonly userProperties?;
    constructor(parent: cdk.Construct, name: string, props: CustomResourceProps);
    /**
     * Override renderProperties to mix in the user-defined properties
     */
    protected renderProperties(): {
        [key: string]: any;
    };
}
