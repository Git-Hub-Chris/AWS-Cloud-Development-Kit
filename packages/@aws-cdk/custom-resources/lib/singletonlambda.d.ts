import iam = require('@aws-cdk/aws-iam');
import lambda = require('@aws-cdk/aws-lambda');
import cdk = require('@aws-cdk/cdk');
/**
 * Properties for a newly created singleton Lambda
 */
export interface SingletonLambdaProps extends lambda.LambdaProps {
    /**
     * A unique identifier to identify this lambda
     *
     * The identifier should be unique across all custom resource providers.
     * We recommend generating a UUID per provider.
     */
    uuid: string;
}
/**
 * A Lambda that will only ever be added to a stack once.
 *
 * The lambda is identified using the value of 'uuid'. Run 'uuidgen'
 * for every SingletonLambda you create.
 */
export declare class SingletonLambda extends lambda.LambdaRef {
    readonly functionName: lambda.FunctionName;
    readonly functionArn: lambda.FunctionArn;
    readonly role?: iam.Role | undefined;
    protected readonly canCreatePermissions: boolean;
    private lambdaFunction;
    constructor(parent: cdk.Construct, name: string, props: SingletonLambdaProps);
    addPermission(name: string, permission: lambda.LambdaPermission): void;
    private ensureLambda;
}
