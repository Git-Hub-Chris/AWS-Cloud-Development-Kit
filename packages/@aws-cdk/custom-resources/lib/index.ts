export * from './aws-custom-resource';
export * from './provider-framework';
export * from './state-machine-provider';
