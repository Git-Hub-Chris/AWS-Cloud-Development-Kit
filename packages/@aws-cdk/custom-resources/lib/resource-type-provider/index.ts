export * from './cfncli-layer';
export * from './resource-type-provider';
