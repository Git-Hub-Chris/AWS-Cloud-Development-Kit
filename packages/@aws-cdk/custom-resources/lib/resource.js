"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_cloudformation_1 = require("@aws-cdk/aws-cloudformation");
/**
 * Custom resource that is implemented using a Lambda
 *
 * As a custom resource author, you should be publishing a subclass of this class
 * that hides the choice of provider, and accepts a strongly-typed properties
 * object with the properties your provider accepts.
 */
class CustomResource extends aws_cloudformation_1.cloudformation.CustomResource {
    constructor(parent, name, props) {
        if (!!props.lambdaProvider === !!props.topicProvider) {
            throw new Error('Exactly one of "lambdaProvider" or "topicProvider" must be set.');
        }
        super(parent, name, {
            serviceToken: props.lambdaProvider ? props.lambdaProvider.functionArn : props.topicProvider.topicArn
        });
        this.userProperties = props.properties;
    }
    /**
     * Override renderProperties to mix in the user-defined properties
     */
    renderProperties() {
        const props = super.renderProperties();
        return Object.assign(props, uppercaseProperties(this.userProperties || {}));
    }
}
exports.CustomResource = CustomResource;
/**
 * Uppercase the first letter of every property name
 *
 * It's customary for CloudFormation properties to start with capitals, and our
 * properties to start with lowercase, so this function translates from one
 * to the other
 */
function uppercaseProperties(props) {
    const ret = {};
    Object.keys(props).forEach(key => {
        const upper = key.substr(0, 1).toUpperCase() + key.substr(1);
        ret[upper] = props[key];
    });
    return ret;
}
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicmVzb3VyY2UuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJyZXNvdXJjZS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUFBLG9FQUE2RDtBQW9DN0Q7Ozs7OztHQU1HO0FBQ0gsb0JBQTRCLFNBQVEsbUNBQWMsQ0FBQyxjQUFjO0lBTTdELFlBQVksTUFBcUIsRUFBRSxJQUFZLEVBQUUsS0FBMEI7UUFDdkUsSUFBSSxDQUFDLENBQUMsS0FBSyxDQUFDLGNBQWMsS0FBSyxDQUFDLENBQUMsS0FBSyxDQUFDLGFBQWEsRUFBRTtZQUNsRCxNQUFNLElBQUksS0FBSyxDQUFDLGlFQUFpRSxDQUFDLENBQUM7U0FDdEY7UUFFRCxLQUFLLENBQUMsTUFBTSxFQUFFLElBQUksRUFBRTtZQUNoQixZQUFZLEVBQUUsS0FBSyxDQUFDLGNBQWMsQ0FBQyxDQUFDLENBQUMsS0FBSyxDQUFDLGNBQWMsQ0FBQyxXQUFXLENBQUMsQ0FBQyxDQUFDLEtBQUssQ0FBQyxhQUFjLENBQUMsUUFBUTtTQUN4RyxDQUFDLENBQUM7UUFFSCxJQUFJLENBQUMsY0FBYyxHQUFHLEtBQUssQ0FBQyxVQUFVLENBQUM7SUFDM0MsQ0FBQztJQUVEOztPQUVHO0lBQ08sZ0JBQWdCO1FBQ3RCLE1BQU0sS0FBSyxHQUFHLEtBQUssQ0FBQyxnQkFBZ0IsRUFBRSxDQUFDO1FBQ3ZDLE9BQU8sTUFBTSxDQUFDLE1BQU0sQ0FBQyxLQUFLLEVBQUUsbUJBQW1CLENBQUMsSUFBSSxDQUFDLGNBQWMsSUFBSSxFQUFFLENBQUMsQ0FBQyxDQUFDO0lBQ2hGLENBQUM7Q0FFSjtBQTFCRCx3Q0EwQkM7QUFFRDs7Ozs7O0dBTUc7QUFDSCw2QkFBNkIsS0FBaUI7SUFDMUMsTUFBTSxHQUFHLEdBQWUsRUFBRSxDQUFDO0lBQzNCLE1BQU0sQ0FBQyxJQUFJLENBQUMsS0FBSyxDQUFDLENBQUMsT0FBTyxDQUFDLEdBQUcsQ0FBQyxFQUFFO1FBQzdCLE1BQU0sS0FBSyxHQUFHLEdBQUcsQ0FBQyxNQUFNLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxDQUFDLFdBQVcsRUFBRSxHQUFHLEdBQUcsQ0FBQyxNQUFNLENBQUMsQ0FBQyxDQUFDLENBQUM7UUFDN0QsR0FBRyxDQUFDLEtBQUssQ0FBQyxHQUFHLEtBQUssQ0FBQyxHQUFHLENBQUMsQ0FBQztJQUM1QixDQUFDLENBQUMsQ0FBQztJQUNILE9BQU8sR0FBRyxDQUFDO0FBQ2YsQ0FBQyJ9