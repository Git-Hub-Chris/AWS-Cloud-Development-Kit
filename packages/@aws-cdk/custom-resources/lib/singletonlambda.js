"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lambda = require("@aws-cdk/aws-lambda");
const cdk = require("@aws-cdk/cdk");
/**
 * A Lambda that will only ever be added to a stack once.
 *
 * The lambda is identified using the value of 'uuid'. Run 'uuidgen'
 * for every SingletonLambda you create.
 */
class SingletonLambda extends lambda.LambdaRef {
    constructor(parent, name, props) {
        super(parent, name);
        this.lambdaFunction = this.ensureLambda(props.uuid, props);
        this.functionArn = this.lambdaFunction.functionArn;
        this.functionName = this.lambdaFunction.functionName;
        this.role = this.lambdaFunction.role;
        this.canCreatePermissions = true; // Doesn't matter, addPermission is overriden anyway
    }
    addPermission(name, permission) {
        return this.lambdaFunction.addPermission(name, permission);
    }
    ensureLambda(uuid, props) {
        const constructName = 'SingletonLambda' + slugify(uuid);
        const stack = cdk.Stack.find(this);
        const existing = stack.tryFindChild(constructName);
        if (existing) {
            // Just assume this is true
            return existing;
        }
        return new lambda.Lambda(stack, constructName, props);
    }
}
exports.SingletonLambda = SingletonLambda;
function slugify(x) {
    return x.replace(/[^a-zA-Z0-9]/g, '');
}
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoic2luZ2xldG9ubGFtYmRhLmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsic2luZ2xldG9ubGFtYmRhLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBQ0EsOENBQStDO0FBQy9DLG9DQUFxQztBQWVyQzs7Ozs7R0FLRztBQUNILHFCQUE2QixTQUFRLE1BQU0sQ0FBQyxTQUFTO0lBT2pELFlBQVksTUFBcUIsRUFBRSxJQUFZLEVBQUUsS0FBMkI7UUFDeEUsS0FBSyxDQUFDLE1BQU0sRUFBRSxJQUFJLENBQUMsQ0FBQztRQUVwQixJQUFJLENBQUMsY0FBYyxHQUFHLElBQUksQ0FBQyxZQUFZLENBQUMsS0FBSyxDQUFDLElBQUksRUFBRSxLQUFLLENBQUMsQ0FBQztRQUUzRCxJQUFJLENBQUMsV0FBVyxHQUFHLElBQUksQ0FBQyxjQUFjLENBQUMsV0FBVyxDQUFDO1FBQ25ELElBQUksQ0FBQyxZQUFZLEdBQUcsSUFBSSxDQUFDLGNBQWMsQ0FBQyxZQUFZLENBQUM7UUFDckQsSUFBSSxDQUFDLElBQUksR0FBRyxJQUFJLENBQUMsY0FBYyxDQUFDLElBQUksQ0FBQztRQUVyQyxJQUFJLENBQUMsb0JBQW9CLEdBQUcsSUFBSSxDQUFDLENBQUMsb0RBQW9EO0lBQzFGLENBQUM7SUFFTSxhQUFhLENBQUMsSUFBWSxFQUFFLFVBQW1DO1FBQ2xFLE9BQU8sSUFBSSxDQUFDLGNBQWMsQ0FBQyxhQUFhLENBQUMsSUFBSSxFQUFFLFVBQVUsQ0FBQyxDQUFDO0lBQy9ELENBQUM7SUFFTyxZQUFZLENBQUMsSUFBWSxFQUFFLEtBQXlCO1FBQ3hELE1BQU0sYUFBYSxHQUFHLGlCQUFpQixHQUFHLE9BQU8sQ0FBQyxJQUFJLENBQUMsQ0FBQztRQUN4RCxNQUFNLEtBQUssR0FBRyxHQUFHLENBQUMsS0FBSyxDQUFDLElBQUksQ0FBQyxJQUFJLENBQUMsQ0FBQztRQUNuQyxNQUFNLFFBQVEsR0FBRyxLQUFLLENBQUMsWUFBWSxDQUFDLGFBQWEsQ0FBQyxDQUFDO1FBQ25ELElBQUksUUFBUSxFQUFFO1lBQ1YsMkJBQTJCO1lBQzNCLE9BQU8sUUFBNEIsQ0FBQztTQUN2QztRQUVELE9BQU8sSUFBSSxNQUFNLENBQUMsTUFBTSxDQUFDLEtBQUssRUFBRSxhQUFhLEVBQUUsS0FBSyxDQUFDLENBQUM7SUFDMUQsQ0FBQztDQUNKO0FBbENELDBDQWtDQztBQUVELGlCQUFpQixDQUFTO0lBQ3RCLE9BQU8sQ0FBQyxDQUFDLE9BQU8sQ0FBQyxlQUFlLEVBQUUsRUFBRSxDQUFDLENBQUM7QUFDMUMsQ0FBQyJ9