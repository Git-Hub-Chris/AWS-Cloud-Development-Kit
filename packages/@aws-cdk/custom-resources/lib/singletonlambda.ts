import { Construct, Stack } from "@aws-cdk/core";
import { Role } from "@aws-cdk/iam";
import { Function, FunctionName, FunctionPermission, FunctionProps, FunctionRef } from "@aws-cdk/lambda";
import { lambda } from "@aws-cdk/resources";

/**
 * Properties for a newly created singleton Lambda
 */
export interface SingletonLambdaProps extends FunctionProps {
    /**
     * A unique identifier to identify this lambda
     *
     * The identifier should be unique across all custom resource providers.
     * We recommend generating a UUID per provider.
     */
    uuid: string;
}

/**
 * A Lambda that will only ever be added to a stack once.
 *
 * The lambda is identified using the value of 'uuid'. Run 'uuidgen'
 * for every SingletonLambda you create.
 */
export class SingletonLambda extends FunctionRef {
    public functionName: FunctionName;
    public functionArn: lambda.FunctionArn;
    public role?: Role | undefined;
    protected canCreatePermissions: boolean;
    private lambdaFunction: FunctionRef;

    constructor(parent: Construct, name: string, props: SingletonLambdaProps) {
        super(parent, name);

        this.lambdaFunction = this.ensureFunction(props.uuid, props);

        this.functionArn = this.lambdaFunction.functionArn;
        this.functionName = this.lambdaFunction.functionName;
        this.role = this.lambdaFunction.role;

        this.canCreatePermissions = true; // Doesn't matter, addPermission is overriden anyway
    }

    public addPermission(name: string, permission: FunctionPermission) {
        return this.lambdaFunction.addPermission(name, permission);
    }

    private ensureFunction(uuid: string, props: FunctionProps): FunctionRef {
        const constructName = 'SingletonLambda' + slugify(uuid);
        const stack = Stack.find(this);
        const existing = stack.tryFindChild(constructName);
        if (existing) {
            // Just assume this is true
            return existing as FunctionRef;
        }

        return new Function(stack, constructName, props);
    }
}

function slugify(x: string): string {
    return x.replace(/[^a-zA-Z0-9]/g, '');
}
