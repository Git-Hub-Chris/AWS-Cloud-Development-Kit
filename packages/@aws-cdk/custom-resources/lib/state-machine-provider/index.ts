export * from './state-machine-provider';
