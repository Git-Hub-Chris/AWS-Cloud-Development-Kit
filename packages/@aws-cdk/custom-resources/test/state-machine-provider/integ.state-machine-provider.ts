import * as iam from '@aws-cdk/aws-iam';
import { App, CfnOutput, CfnResource, Construct, CustomResource, Stack } from '@aws-cdk/core';
import { IStateMachine, StateMachineProvider } from '../../lib';

/*
 * Stack verification steps:
 * * aws cloudformation describe-stacks --stack-name cdk-integ-state-machine-provider --query Stacks[0].Outputs[0].OutputValue
 *   should return "bar"
 */
class TestStack extends Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const provider = new StateMachineProvider(this, 'Provider', {
      stateMachine: new PassStateMachine(this, 'PassStateMachine', {
        result: {
          PhysicalResourceId: 'physical-resource-id',
          Data: {
            foo: 'bar',
          },
        },
      }),
    });

    const cr = new CustomResource(this, 'CustomResource', {
      serviceToken: provider.serviceToken,
      properties: {
        Key: 'value',
      },
    });

    new CfnOutput(this, 'Foo', {
      value: cr.getAttString('foo'),
    });
  }
}

interface PassStateMachineProps {
  result: any;
}

class PassStateMachine extends Construct implements IStateMachine {
  public readonly stateMachineArn: string;

  constructor(scope: Construct, id: string, props: PassStateMachineProps) {
    super(scope, id);

    const role = new iam.Role(this, 'Role', {
      assumedBy: new iam.ServicePrincipal('states.amazonaws.com'),
    });

    const definition = Stack.of(this).toJsonString({
      StartAt: 'Pass',
      States: {
        Pass: {
          Type: 'Pass',
          Result: props.result,
          End: true,
        },
      },
    });

    const passStateMachine = new CfnResource(this, 'Resource', {
      type: 'AWS::StepFunctions::StateMachine',
      properties: {
        DefinitionString: definition,
        RoleArn: role.roleArn,
      },
    });
    passStateMachine.node.addDependency(role);
    this.stateMachineArn = passStateMachine.ref;
  }
}

const app = new App();

new TestStack(app, 'cdk-integ-state-machine-provider');

app.synth();
