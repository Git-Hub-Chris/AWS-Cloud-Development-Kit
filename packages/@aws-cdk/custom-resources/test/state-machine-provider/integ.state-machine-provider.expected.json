{
  "Resources": {
    "PassStateMachineRoleFFB3319B": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "states.",
                      {
                        "Ref": "AWS::Region"
                      },
                      ".amazonaws.com"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PassStateMachine597FE51C": {
      "Type": "AWS::StepFunctions::StateMachine",
      "Properties": {
        "DefinitionString": "{\n  \"StartAt\": \"Pass\",\n  \"States\": {\n    \"Pass\": {\n      \"Type\": \"Pass\",\n      \"Result\": {\n        \"PhysicalResourceId\": \"physical-resource-id\",\n        \"Data\": {\n          \"foo\": \"bar\"\n        }\n      },\n      \"End\": true\n    }\n  }\n}",
        "RoleArn": {
          "Fn::GetAtt": [
            "PassStateMachineRoleFFB3319B",
            "Arn"
          ]
        }
      },
      "DependsOn": [
        "PassStateMachineRoleFFB3319B"
      ]
    },
    "ProviderCfnResponseSuccessServiceRole5D01681D": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "ProviderCfnResponseSuccess01B6851D": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3BucketF2E5E48C"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.cfnResponseSuccess",
        "Role": {
          "Fn::GetAtt": [
            "ProviderCfnResponseSuccessServiceRole5D01681D",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "ProviderCfnResponseSuccessServiceRole5D01681D"
      ]
    },
    "ProviderCfnResponseFailedServiceRole1026DD92": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "ProviderCfnResponseFailedCC9A90F2": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3BucketF2E5E48C"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.cfnResponseFailed",
        "Role": {
          "Fn::GetAtt": [
            "ProviderCfnResponseFailedServiceRole1026DD92",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "ProviderCfnResponseFailedServiceRole1026DD92"
      ]
    },
    "ProviderRole1DE8EC3B": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "states.",
                      {
                        "Ref": "AWS::Region"
                      },
                      ".amazonaws.com"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "ProviderRoleDefaultPolicyD21592EE": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ProviderCfnResponseSuccess01B6851D",
                    "Arn"
                  ]
                },
                {
                  "Fn::GetAtt": [
                    "ProviderCfnResponseFailedCC9A90F2",
                    "Arn"
                  ]
                }
              ]
            },
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "PassStateMachine597FE51C"
              }
            },
            {
              "Action": [
                "states:DescribeExecution",
                "states:StopExecution"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":states:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":execution:",
                    {
                      "Fn::Select": [
                        6,
                        {
                          "Fn::Split": [
                            ":",
                            {
                              "Ref": "PassStateMachine597FE51C"
                            }
                          ]
                        }
                      ]
                    },
                    "*"
                  ]
                ]
              }
            },
            {
              "Action": [
                "events:PutTargets",
                "events:PutRule",
                "events:DescribeRule"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":events:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":rule/StepFunctionsGetEventsForStepFunctionsExecutionRule"
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "ProviderRoleDefaultPolicyD21592EE",
        "Roles": [
          {
            "Ref": "ProviderRole1DE8EC3B"
          }
        ]
      }
    },
    "ProviderStateMachine873099F5": {
      "Type": "AWS::StepFunctions::StateMachine",
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{\n  \"StartAt\": \"StartExecution\",\n  \"States\": {\n    \"StartExecution\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"arn:aws:states:::states:startExecution.sync:2\",\n      \"Parameters\": {\n        \"Input.$\": \"$\",\n        \"StateMachineArn\": \"",
              {
                "Ref": "PassStateMachine597FE51C"
              },
              "\"\n      },\n      \"TimeoutSeconds\": 1800,\n      \"Next\": \"CfnResponseSuccess\",\n      \"Catch\": [\n        {\n          \"ErrorEquals\": [\n            \"States.ALL\"\n          ],\n          \"Next\": \"CfnResponseFailed\"\n        }\n      ]\n    },\n    \"CfnResponseSuccess\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"",
              {
                "Fn::GetAtt": [
                  "ProviderCfnResponseSuccess01B6851D",
                  "Arn"
                ]
              },
              "\",\n      \"End\": true\n    },\n    \"CfnResponseFailed\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"",
              {
                "Fn::GetAtt": [
                  "ProviderCfnResponseFailedCC9A90F2",
                  "Arn"
                ]
              },
              "\",\n      \"End\": true\n    }\n  }\n}"
            ]
          ]
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "ProviderRole1DE8EC3B",
            "Arn"
          ]
        }
      },
      "DependsOn": [
        "ProviderRoleDefaultPolicyD21592EE",
        "ProviderRole1DE8EC3B"
      ]
    },
    "ProviderStartExecutionServiceRole86F1DDB6": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "ProviderStartExecutionServiceRoleDefaultPolicy9428BBFE": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "ProviderStateMachine873099F5"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "ProviderStartExecutionServiceRoleDefaultPolicy9428BBFE",
        "Roles": [
          {
            "Ref": "ProviderStartExecutionServiceRole86F1DDB6"
          }
        ]
      }
    },
    "ProviderStartExecutionA94D3483": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3BucketF2E5E48C"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.startExecution",
        "Role": {
          "Fn::GetAtt": [
            "ProviderStartExecutionServiceRole86F1DDB6",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x",
        "Environment": {
          "Variables": {
            "STATE_MACHINE_ARN": {
              "Ref": "ProviderStateMachine873099F5"
            }
          }
        }
      },
      "DependsOn": [
        "ProviderStartExecutionServiceRoleDefaultPolicy9428BBFE",
        "ProviderStartExecutionServiceRole86F1DDB6"
      ]
    },
    "CustomResource": {
      "Type": "AWS::CloudFormation::CustomResource",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "ProviderStartExecutionA94D3483",
            "Arn"
          ]
        },
        "Key": "value"
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    }
  },
  "Parameters": {
    "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3BucketF2E5E48C": {
      "Type": "String",
      "Description": "S3 bucket for asset \"4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2\""
    },
    "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2S3VersionKey3870C770": {
      "Type": "String",
      "Description": "S3 key for asset version \"4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2\""
    },
    "AssetParameters4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2ArtifactHashE4E90A37": {
      "Type": "String",
      "Description": "Artifact hash for asset \"4b54c70e5567da65d43f00f105bd1ea3e26f2331f8a897a0dd4e8b071a9adca2\""
    }
  },
  "Outputs": {
    "Foo": {
      "Value": {
        "Fn::GetAtt": [
          "CustomResource",
          "foo"
        ]
      }
    }
  }
}